### ISSUE TITLE
Unused local variable reassignment with the same name is not reported (F841?)

### ISSUE THREAD

**Issue by MiroPsota** [astral-sh/ruff#14355]
In the following code snippet, the first assignment to `a` is never used, but is not reported by Ruff 0.7.3. PyCharm reports it and suggests to delete assignment target.
```
def main():
    a = 1
    a = 2
    print(a)
```

**Comment by dylwil3** [comment]
Thanks, this is interesting! Unless I'm misusing the tool, it doesn't seem like Pyflakes flags this either. I'm wondering if `F841` is restricted in scope to literally not using the _name_ of some variable.

I also tried it in `pyright` and `mypy` to see if a type checker would complain and didn't see any issues.

```console
➜  cat tmp.py
def main():
    a = 1
    a = 2
    print(a)%                                                                           
➜  uvx pyflakes tmp.py
➜  
```
Is the suggestion that we add a lint rule for this? Do you happen to know of a lint rule in one of the (partially) supported plugins that catches this, or is it just something Pycharm does?

**Comment by charliermarsh** [comment]
Yeah this matches the Pyflakes behavior, which is intentional IIRC.

**Comment by MiroPsota** [comment]
If it's intentional, then I suggest making to make a new rule, as it's useful IMO.

**Comment by MichaReiser** [comment]
The relevant upstream issue is https://github.com/PyCQA/pyflakes/issues/498. So this is more of a technical limitation than it isn't in the rule's intent. Ruff might be able to detect this very trivial case (if all assignments are all in the same branch).

This is the same underlying issue as https://github.com/astral-sh/ruff/issues/6704
